zerver/tests/test_messages.py

```diff
@@ -2731,3 +2733,38 @@ def test_messages_for_ids(self) -> None:
 
         self.assertIn('class="user-mention"', new_message['content'])
         self.assertEqual(new_message['flags'], ['mentioned'])
+
+class MessageVisibilityTest(ZulipTestCase):
+    def test_get_first_visible_message_id_with_no_limit(self) -> None:
+        realm = get_realm("zulip")
+        realm.message_visibility_limit = None
+        realm.save()
+        cache_delete(realm_last_visible_message_id_cache_key(realm))
+
+        self.assertEqual(get_first_visible_message_id(realm), 0)
+
+    def test_get_first_visible_message_id_with_limit(self) -> None:
+        for i in range(5):
+            self.send_stream_message(self.example_email("othello"), "Scotland")
+
+        realm = get_realm("zulip")
+        num_messages = Message.objects.filter(sender__realm=realm).count()
+        # Show only the last (num_messages - 4) messages
+        message_visibility_limit = num_messages - 4
+        realm.message_visibility_limit = message_visibility_limit
+        realm.save()
+
+        cache_delete(realm_last_visible_message_id_cache_key(realm))
+
+        expected_id = Message.objects.filter(sender__realm=realm).values('id').\
+            order_by('-id')[message_visibility_limit - 1]["id"]
```

Comment:
This test makes me nervous, because the way you set the expected value is **so** close to the way you implement it, there's no real guarantee it will fail if the implementation is slightly broken.

I'd do something like this:

~~~ py
Message.objects.delete()
message_ids = [self.send_stream_message(self.example_email("othello"), "Scotland").id for i in range(200)]
realm.message_visibility_limit = 130
expected_message_id = message_ids[70]
# ...